client/modules/components/src/Wikitext.svelte

```diff
@@ -0,0 +1,147 @@
+<!--
+  @component Generic wikitext container.
+-->
+<script lang="ts">
+  import * as FTML from "ftml-wasm-worker"
+  import { createAnimQueued, createMutatingLock, perfy, toFragment } from "wj-util"
+  import Card from "./Card.svelte"
+  import morphdom from "morphdom"
+  import Spinny from "./Spinny.svelte"
+  import { anim } from "./lib/animation"
+
+  type Rendered = { html: string; styles: string[] }
+  type WikitextInput =
+    | Promisable<string>
+    | Promisable<Rendered>
+    | (() => Promisable<string>)
+    | (() => Promisable<Rendered>)
+
+  /**
+   * Wikitext input. May be given in quite a few ways:
+   * 1. As a string.
+   * 2. As an object, `{ html: string, style: string }`.
+   * 3. As a function returning either a string or the aformentioned object.
+   * 4. As a `Promise` resolving to either a string or the aforementioned object.
+   * 5. As a function returning a `Promise` like the aformentioned one.
+   *
+   * Providing the input as a function has performance benefits. If the wikitext
+   * takes a long time to render, and the source wikitext is frequently updating,
+   * providing a function will make it so that the wikitext is only evalulated when
+   * it needs to be. Otherwise, the wikitext would be evalulated constantly, and that
+   * has obvious performance implications.
+   */
+  export let wikitext: WikitextInput = ""
+
+  /**
+   * Flags the component to morph the output container
+   * rather than entirely replacing the contents.
+   *
+   * This is for performance reasons. If you are frequently updating the wikitext,
+   * it's probably best if the DOM is morphed rather than replaced.
+   */
+  export let morph = false
+
+  let element: HTMLElement
+  let stylesheet: HTMLStyleElement
+
+  let rendering = false
+
+  let perfRender = 0
+
+  const render = createMutatingLock(async (wikitext: WikitextInput) => {
+    const displayIndicatorTimeout = setTimeout(() => (rendering = true), 100)
+    const measure = perfy()
+    if (typeof wikitext === "function") wikitext = wikitext()
+    wikitext = await wikitext
+    const result: Rendered = wikitext
+      ? typeof wikitext !== "string"
+        ? wikitext
+        : await FTML.render(wikitext)
+      : { html: "", styles: [""] }
+    perfRender = measure()
+    clearTimeout(displayIndicatorTimeout)
+    return result
+  })
+
+  // TODO: Security audit of this - how much should we trust FTML output right now?
```

Comment:
Well, in the early stages if Wikijump gets used anywhere it may be a good idea to not trust FTML until we get more time with it.

If the code was fully untrusted, I'd probably hook up DOMPurify to it, along with a bunch of rules/hooks so that DOMPurify doesn't sanitize out some normally blacklisted syntax.

Reply:
Well at least note that for rendered page output, PHP puts any output through an HTML purifier, so at least those can be trusted.

Reply:
Oh also I think with DOMPurify I could hook into it and have it report "violations", which may be interesting to log. I'm not sure how often it's just going to report minor nitpicks or not but I could play with it

Reply:
For right now I'm going to trust FTML's output - it's locked down pretty well I think. We can maybe hedge our bets that the browser native sanitizer API will be somewhat available by the time we start getting users, and we could use that. Or just be cautious, idk.